Camera RAW files picked via file input are returned as PNG on change

pxlcoder opened the pull request:
#### 00c3f53f2de429d9828d9e065dc989d0063b6fc8
<pre>
Camera RAW files picked via file input are returned as PNG on change
<a href="https://bugs.webkit.org/show_bug.cgi?id=258467">https://bugs.webkit.org/show_bug.cgi?id=258467</a>
rdar://111231838

Reviewed by Cameron McCormack.

Camera RAW type identifiers are registered with the system via the RawCamera
bundle. While the list of supported RAW types is fairly comprehensive, many
type identifiers are missing an associated MIME type.

Consequently, when choosing a RAW file (such as NEF) for an input that supports
RAW files and PNGs, the lack of correct information in the RawCamera bundle
results in &quot;.nef&quot; getting mapped to a nil MIME type.

WebKit also contains logic to automatically transcode images selected in file
inputs to an accepted MIME type, if the current MIME type is unsupported, and
the file is an image type. For file inputs that specify `accept` values in the
form of extensions, the extensions are normalized to MIME types.

For an input that supports RAW files and PNGs, WebKit attempts to transcode
the image, as &quot;image/png&quot; is the only recognized MIME type, following the system
lookup. Transcoding then succeeds, since ImageIO knows about the RAW type, and
is able to convert it to a PNG.

To fix, workaround the missing MIME type mappings in RawCamera.bundle, by
explicitly declaring them in WebKit.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/file/entries-api/png-raw-open-panel-expected.txt: Added.
* LayoutTests/fast/forms/file/entries-api/png-raw-open-panel.html: Added.
* LayoutTests/fast/forms/file/entries-api/resources/images/image.nef: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
(WebCore::additionalMimeTypesMap):
(WebCore::additionalExtensionsMap):
(WebCore::MIMETypeRegistry::mimeTypeForExtension):
(WebCore::MIMETypeRegistry::extensionsForMIMEType):
(WebCore::MIMETypeRegistry::preferredExtensionForMIMEType):

Canonical link: <a href="https://commits.webkit.org/266049@main">https://commits.webkit.org/266049@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fb035008941b6eef2446dfb2c891350f918bd45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12675 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13005 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14414 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12130 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12736 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13023 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14832 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10726 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14857 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10884 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11472 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18558 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11638 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14841 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/12115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10028 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11366 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11343 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3112 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/15693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -78,6 +78,58 @@
+            { "image/x-samsung-srw"_s, "srw"_s },
+            { "image/x-sony-arw"_s, "arw"_s },
+            { "image/x-sony-srf"_s, "srf"_s },
+
```
Nit: drop the blank line.

heycam reviewed (approved):
r=me but I did not review the specific set of types / extensions.

aprotyas commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -78,6 +78,57 @@
+        };
+        for (auto& pair : additionalTypes) {
+            ASCIILiteral type = pair.type;
+            ASCIILiteral extension = pair.extension;
```
Nit: since `TypeExtensionPair` is a simple struct, I think you can use a structured binding declaration here?

```cpp
for (auto& [type, extension] : additionalTypes)
    map.add(extension, type);
```

pxlcoder replied:
Done!

aprotyas commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -78,6 +78,57 @@
+// Specify MIME type <-> extension mappings for type identifiers recognized by the system that are missing MIME type values.
+static const HashMap<String, String, ASCIICaseInsensitiveHash>& additionalMimeTypesMap()
+{
+    ASSERT(isMainThread());
```
For my own education, why do we want this assertion?

pxlcoder replied:
I had copied this from somewhere else with similar logic. I've removed it now, since it isn't really necessary here.

aprotyas commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -92,14 +143,32 @@
+    if (mimeType.length)
+        return mimeType;
+
+    auto mapEntry = additionalMimeTypesMap().find<ASCIICaseInsensitiveStringViewHashTranslator>(extension);
+    if (mapEntry != additionalMimeTypesMap().end())
```
Nit

```suggestion
    if (auto mapEntry = additionalMimeTypesMap().find<ASCIICaseInsensitiveStringViewHashTranslator>(extension); mapEntry != additionalMimeTypesMap().end())
```

aprotyas commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -92,14 +143,32 @@
         return extensionsForWildcardMIMEType(type);
-    return makeVector<String>([[NSURLFileTypeMappings sharedMappings] extensionsForMIMEType:type]);
+
+    NSArray *extensions = [[NSURLFileTypeMappings sharedMappings] extensionsForMIMEType:type];
+    if (extensions.count)
```
Very optional nit suggestion to initialize the variable of interest in init statements for all the `if` expressions introduced in this PR, but ignore if you think it harms readability.

pxlcoder replied:
Going to keep this as-is, unless you feel strongly, since I generally only use that pattern if the `if` expression is just a `nil`/`NULL` check. 

I think it is slightly more readable, avoiding a longer line of code, and makes it easier to add debug logging when needed.

aprotyas replied:
Sounds good, not a strong preference

pxlcoder committed:
Camera RAW files picked via file input are returned as PNG on change
https://bugs.webkit.org/show_bug.cgi?id=258467
rdar://111231838

Reviewed by Cameron McCormack.

Camera RAW type identifiers are registered with the system via the RawCamera
bundle. While the list of supported RAW types is fairly comprehensive, many
type identifiers are missing an associated MIME type.

Consequently, when choosing a RAW file (such as NEF) for an input that supports
RAW files and PNGs, the lack of correct information in the RawCamera bundle
results in ".nef" getting mapped to a nil MIME type.

WebKit also contains logic to automatically transcode images selected in file
inputs to an accepted MIME type, if the current MIME type is unsupported, and
the file is an image type. For file inputs that specify `accept` values in the
form of extensions, the extensions are normalized to MIME types.

For an input that supports RAW files and PNGs, WebKit attempts to transcode
the image, as "image/png" is the only recognized MIME type, following the system
lookup. Transcoding then succeeds, since ImageIO knows about the RAW type, and
is able to convert it to a PNG.

To fix, workaround the missing MIME type mappings in RawCamera.bundle, by
explicitly declaring them in WebKit.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/file/entries-api/png-raw-open-panel-expected.txt: Added.
* LayoutTests/fast/forms/file/entries-api/png-raw-open-panel.html: Added.
* LayoutTests/fast/forms/file/entries-api/resources/images/image.nef: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
(WebCore::additionalMimeTypesMap):
(WebCore::additionalExtensionsMap):
(WebCore::MIMETypeRegistry::mimeTypeForExtension):
(WebCore::MIMETypeRegistry::extensionsForMIMEType):
(WebCore::MIMETypeRegistry::preferredExtensionForMIMEType):

Canonical link: https://commits.webkit.org/266049@main
